metricbeat/module/windows/wmi/wmi.go

```diff
@@ -0,0 +1,212 @@
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The windows wmi metricset is beta.")
+
+	config := NewDefaultConfig()
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	err := config.ValidateConnectionParameters()
+	if err != nil {
+		return nil, err
+	}
+
+	err = config.CompileQueries()
+	if err != nil {
+		return nil, err
+	}
+
+	err = config.ApplyDefaultNamespaceToQueries(config.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	config.BuildNamespaceQueryIndex()
+
+	if config.WarningThreshold == 0*time.Second {
+		config.WarningThreshold = base.Module().Config().Period
+	}
+
+	m := &MetricSet{
+		BaseMetricSet: base,
+		config:        config,
+	}
+
+	return m, nil
+}
+
+// This function handles the skip conditions
+func (m *MetricSet) shouldSkipNilOrEmptyValue(fieldValue interface{}) bool {
+	if fieldValue == nil {
+		if !m.config.IncludeNull {
+			return true // Skip if it's nil and IncludeNull is false
+		}
+	} else if stringValue, ok := fieldValue.(string); ok {
+		if len(stringValue) == 0 && !m.config.IncludeEmptyString {
+			return true // Skip if it's an empty string and IncludeEmptyString is false
+		}
+	}
+	return false
+}
+
+// Fetch method implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) error {
+
+	sm := wmi.NewWmiSessionManager()
+	defer sm.Dispose()
+
+	// To optimize performance and reduce overhead, we create a single session
+	// for each unique WMI namespace. This minimizes the number of session creations
+	for namespace, queries := range m.config.NamespaceQueryIndex {
+
+		session, err := sm.GetSession(namespace, m.config.Host, "", m.config.User, m.config.Password)
+
+		if err != nil {
+			return fmt.Errorf("could not initialize session %w", err)
+		}
+		_, err = session.Connect()
+		if err != nil {
+			return fmt.Errorf("could not connect session %w", err)
+		}
+		defer session.Dispose()
+
+		// We create a shared conversion table for entries with the same schema.
+		// This avoids repeatedly fetching the schema for each individual entry, improving efficiency.
+		conversionTable := make(map[string]WmiStringConversionFunction)
+
+		for _, queryConfig := range queries {
+
+			query := queryConfig.QueryStr
+
+			rows, err := ExecuteGuardedQueryInstances(session, query, m.config.WarningThreshold, m.Logger())
+
+			if err != nil {
+				m.Logger().Warn("Could not execute query: %v", err)
+				continue
+			}
+
+			defer wmi.CloseAllInstances(rows)
+
+			for _, instance := range rows {
+				event := mb.Event{
+					MetricSetFields: mapstr.M{
+						"class":     queryConfig.Class,
+						"namespace": m.config.Namespace,
+						"host":      m.config.Host,
```

Comment:
and i think we should probably leave `host` to the `add_host_metadata` processor, which will use ecs-compliant field names (in this case `host.name`)

Reply:
sorry never mind, the agent can be running on a different host to where the queries are being executed, and `host` is in the metricset field scope, so this is all fine. please ignore.